src/lib/device-logs/backends/loki.ts

```diff
@@ -0,0 +1,194 @@
+					if (messageObj.streams) {
+						this.fromLokiStreamsToDeviceLogs(messageObj.streams).forEach(
+							(log: DeviceLog) => {
+								this.subscriptions.emit(key, log);
+							},
+						);
+					}
+				} catch (err) {
+					captureException(err, 'Unexpected message received from Loki tail');
+				}
+			});
+			ws.on('close', () => {
+				this.webSockets.delete(key);
+			});
+			this.webSockets.set(key, ws);
+		}
+		this.subscriptions.on(key, subscription);
+	}
+
+	public unsubscribe(ctx: LogContext, subscription: Subscription) {
+		const key = this.getKey(ctx);
+		this.subscriptions.removeListener(key, subscription);
+		if (this.webSockets.has(key)) {
+			const ws = this.webSockets.get(key);
+			ws?.close();
+		}
+	}
+
+	private getDeviceQuery(ctx: LogContext) {
+		return `{device_id="${ctx.id}"}`;
+	}
+
+	private getKey(ctx: LogContext, suffix = 'logs') {
+		return `app:${ctx.belongs_to__application}:device:${ctx.id}:${suffix}`;
+	}
+
+	private fromLokiStreamsToDeviceLogs(streams: LokiStream[]): DeviceLog[] {
+		return streams.flatMap(this.fromLokiStream);
+	}
+
+	private fromLokiStream(stream: LokiStream): DeviceLog[] {
+		try {
+			return stream.values.map(([, message]) => {
+				const log = JSON.parse(message);
+				if (log.version !== VERSION) {
+					throw new Error(
+						`Invalid Loki serialization version: ${JSON.stringify(log)}`,
+					);
+				}
+				delete log.version;
+				return log as DeviceLog;
+			});
+		} catch (err) {
+			captureException(err, `Failed to deserialize a Loki log: ${stream}`);
+			return [];
+		}
+	}
+
+	private validateLog(log: DeviceLog) {
+		if (typeof log.message !== 'string') {
+			throw Error('DeviceLog message must be string');
+		} else if (typeof log.timestamp !== 'number') {
+			throw Error('DeviceLog timestamp must be number');
+		} else if (typeof log.isSystem !== 'boolean') {
+			throw Error('DeviceLog isSystem must be boolean');
+		} else if (typeof log.isStdErr !== 'boolean') {
+			throw Error('DeviceLog isStdErr must be boolean');
+		} else if (
+			typeof log.serviceId !== 'number' &&
+			log.serviceId !== undefined
+		) {
+			throw Error('DeviceLog serviceId must be number or undefined');
+		}
+	}
+
+	private toLokiStream(
+		ctx: LogContext,
+		log: DeviceLog & { version?: number },
+	): LokiStream {
+		try {
+			this.validateLog(log);
+			log.version = VERSION;
+			return {
+				stream: {
+					device_id: String(ctx.id), // write only device id, to reduce cardinality do not also write device uuid
+					service_id: log.serviceId ? String(log.serviceId) : 'null',
+					is_system: String(log.isSystem),
+					is_std_err: String(log.isStdErr),
+				},
+				values: [
+					[
+						`${log.timestamp}000000`, // convert timestamp to nanosecond string
+						JSON.stringify(log), // store log as JSON string to retain any meta not writen as label
+					],
+				],
+			};
+		} catch (err) {
+			// Rethrow with a type of error that will end up as status 400
+			throw new BadRequestError(err);
```

Comment:
The only thing that can throw is the `this.validateLog(log);` right? In that case I'd prefer they throw `BadRequestError`s directly and avoid this rewrapping

Reply:
updated